code/lib/builder-vite/src/plugins/externals-plugin.ts

```diff
@@ -0,0 +1,119 @@
+import { join } from 'node:path';
+import { init, parse } from 'es-module-lexer';
+import MagicString from 'magic-string';
+import { emptyDir, ensureDir, ensureFile, writeFile } from 'fs-extra';
+import { mergeAlias } from 'vite';
+import type { Alias, Plugin } from 'vite';
+
+const escapeKeys = (key: string) => key.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&');
+const defaultImportRegExp = 'import ([^*{}]+) from';
+const replacementMap = new Map([
+  ['import ', 'const '],
+  ['import{', 'const {'],
+  ['* as ', ''],
+  [' as ', ': '],
+  [' from ', ' = '],
+  ['}from', '} ='],
+]);
+
+/**
+ * This plugin swaps out imports of pre-bundled storybook preview modules for destructures from global
+ * variables that are added in runtime.mjs.
+ *
+ * For instance:
+ *
+ * ```js
+ * import { useMemo as useMemo2, useEffect as useEffect2 } from "@storybook/preview-api";
+ * ```
+ *
+ * becomes
+ *
+ * ```js
+ * const { useMemo: useMemo2, useEffect: useEffect2 } = __STORYBOOK_MODULE_PREVIEW_API__;
+ * ```
+ *
+ * It is based on existing plugins like https://github.com/crcong/vite-plugin-externals
+ * and https://github.com/eight04/rollup-plugin-external-globals, but simplified to meet our simple needs.
+ */
+export async function externalsPlugin(externals: Record<string, string>) {
```

Comment:
Should probably change this as well along with the filename for uniformity `external-globals-plugin.ts`
```suggestion
export async function externalGlobalsPlugin(externals: Record<string, string>) {
```

Reply:
We can also do this in a followup if we just want to get it merged quickly.  I wouldn't treat it as a blocker, personally.

Reply:
Agreed